lua/luasnip/nodes/snippet.lua

```diff
@@ -47,6 +47,19 @@ local function S(context, nodes, condition, ...)
 	if type(context) == "string" then
 		context = { trig = context }
 	end
+
+	local has_i0 = false
+	for _, node in pairs(nodes) do
+		if node.type == 1 and node.pos == 0 then
+			has_i0 = true
+		end
+	end
+
+	if not has_i0 then
+		local i0 = iNode.I(0)
+		nodes[#nodes + 1] = iNode.I(0)
+	end
+
```

Comment:
I think it is also possible to check for `snip.insert_nodes[0]` after `snip:init_nodes()` and then insert the `iNode.I(0)` in `insert_nodes[0]` and `nodes[#nodes+1]`, which would be a bit cleaner IMO.

Reply:
`snip:init_nodes` is called for all node types, not just snippets, which requires more type-checking, but also this has a narrower scope (it's clearly *just* for snippets here). Why would you find it cleaner in `init_nodes`? There's one more iteration here but in practice it's really short and quick.

Also line 59 should go, sorry about that.

Reply:
Oh, not in `init_nodes`, I meant after the call to `init_nodes`. In there we also step through all nodes and assign them to `snip.insert_nodes`, which means that after it returns, `insert_nodes[0]` contains the `i(0)` (if of exists).

Reply:
We'd then also need to set `i(0).parent` to `self` and `i(0).indx` to 0, but that's better than searching thtough `nodes` two times.

Reply:
I'm not sure that removing a tight loop over a handful of elements is worth the overhead in complexity/readability, but your call!

Reply:
Whoops, the `i0.indx = 0`was wrong, srry about that. `node.indx` actually represents where the node is located, not when it is jumped to :grimacing: